lib/find-package-path.js

```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+var heimdall = require('heimdalljs');
+var path = require('path');
+var IS_WINDOWS = (require('os').platform() === 'win32');
+var getRealFilePath = require('./get-real-file-path');
+
+
+/*
+ * Given a package 'name' and starting directory, resolve to a real (existing) file path.
+ *
+ * Do it similar to what is done in resolve.sync() - travel up the directory hierarchy,
+ * attaching 'node-modules' to each directory and seeing if the directory exists and
+ * has the relevant 'package.json' file we're searching for. It is *much* faster than
+ * resolve.sync(), because we don't test that the requested name is a directory.
+ * This routine assumes that it is only called when we don't already have
+ * the cached entry in packageCache.
+ *
+ * @param name the 'name' of the package, i.e. x in require(x), but with
+ *   '/package.json' on the end. It is NOT a directory (so we don't have to do the
+ *   directory checks that resolve.sync does).
+\ *
+ * @param startPath the directory (non-empty) to start from, appending the name to the
+ *   directory and checking that the file exists. Go up the directory hierarchy from there.
+ *
+ * @result the path to the actual package.json file that's found, or null if not.
+ */
+module.exports = function findPackagePath(realFilePathCache, name, dir) {
+
+  var h = heimdall.start('hashForDep:findPackagePath');
+
+  try {
+    // Early versions of Node don't support 'path.parse', and Windows
+    // doesn't use '/', so here we go.
+    var fsRoot = (IS_WINDOWS ? dir.split(path.sep)[0] : '/');
```

Comment:
Since it's only for getting the root, path-root is better. I thought a straightforward implementation was "better", but agree it's something of a hack. Let me check on path-root in the morning.